src/charm.py

```diff
@@ -390,16 +390,23 @@ def _get_redis_relation_data(self) -> typing.Tuple[str, int]:
         Raises:
             MissingRedisRelationDataError if the some of redis relation data is malformed/missing
         """
-        relation_data = self.redis.relation_data
-        if not relation_data:
+        relation = self.model.get_relation(self.redis.relation_name)
+        if not relation:
             raise MissingRedisRelationDataError("No redis relation data")
+        relation_app_data = relation.data[relation.app]
+        redis_hostname = (
+            str(relation_app_data.get("leader-host"))
+            if relation_app_data.get("leader-host")
+            else None
+        )
 
         try:
-            redis_hostname = str(relation_data["hostname"])
-            redis_port = int(relation_data["port"])
+            if not redis_hostname:
+                redis_hostname = str(self.redis.relation_data["hostname"])
+            redis_port = int(self.redis.relation_data["port"])
         except (KeyError, ValueError) as exc:
             raise MissingRedisRelationDataError(
-                "Wrong hostname or port in Redis relation"
+                "Wrong hostname or port in Redis App data"
             ) from exc
```

Comment:
The whole structure of the code bothers me a bit, can't we do something like:
```python
try:
   redis_port = int(relation_data["port"])
   redis_hostname = str(relation_app_data["leader-host"] if relation_app_data.get("leader-host") else relation_app_data["hostname"])
except (KeyError, ValueError) as exc:
   raise MissingRedisRelationDataError("Either 'leader-host' or 'hostname' and 'ports' are mandatory") from exc
```

Or something around the lines

Reply:
Sure can do! I will also do the same in Indico Charm.